intefaces/builtin, cmd/snap: Kerberos tickets support

nteodosio opened the pull request:
Expose Kerberos tickets in /tmp to snaps via the hostfs.

The snaps will need to connect the kerberos-ticket interface (slot also here provided) to actually be able to read the tickets.

[LP: #1849346](https://bugs.launchpad.net/ubuntu/+source/chromium-browser/+bug/1849346)
[UDENG-7057](https://warthogs.atlassian.net/browse/UDENG-7057)

[UDENG-7057]: https://warthogs.atlassian.net/browse/UDENG-7057?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

zyga commented on cmd/snap/cmd_run.go:
```diff
@@ -760,6 +760,36 @@ func snapdHelperPath(toolName string) (string, error) {
+ * kerberos-tickets interface to have access to the tickets.
+ */
+func exposeKerberosTickets(info *snap.Info) (string, error) {
+	ok := false
+
+	krb5EnvVar := osGetenv("KRB5CCNAME")
+	if len(krb5EnvVar) == 0 {
+		ok = true
+	} else if strings.HasPrefix(krb5EnvVar, "FILE:") {
+		path := strings.TrimPrefix(krb5EnvVar, "FILE:")
+		path = filepath.Clean(path)
+		ok = strings.HasPrefix(path, "/tmp/krb5cc")
+	} else if strings.HasPrefix(krb5EnvVar, "DIR:/tmp") {
+		path := strings.TrimPrefix(krb5EnvVar, "DIR:")
+		path = filepath.Clean(path)
+		ok = path == "/tmp"
+	}
+
+	if ok {
+		return "/var/lib/snapd/hostfs/tmp", nil
+	}
+	return "", fmt.Errorf("Unsupported KRB5CCNAME: %s", krb5EnvVar)
```
This is all okay but it feels very C-like

I'd write it more directly

If env is empty, return "", nil
If env has FILE: prefix, validate and return
If env has DIR: prefix, validate and return
Otherwise return "", error

I'd probably use switch/case for the matching

zyga commented on interfaces/builtin/kerberos_tickets.go:
```diff
@@ -0,0 +1,44 @@
+`
+
+const krbConnectedPlugAppArmor = `
+/var/lib/snapd/hostfs/tmp/krb5cc* r,
```
How does the DIR: case work in this manner? I think we'd need more permissions for it to do something.

Can you provide an example of how real-world DIR: value might look like?

nteodosio replied:
`DIR:/tmp` will cause ticket files to be created in `/tmp`, so access to `/tmp/krb5cc*` would still be enough.

nteodosio replied:
I also had to add the lock permission.

zyga reviewed (commented):
First pass, looks pretty good but I would like to understand the DIR: case better.

nteodosio commented:
Now tested with Firefox in a 24.04 virtual machine.

The interface/plug was emulated by manually inserting the rule in `/var/lib/snapd/apparmor/profiles/snap.firefox.firefox`:

```
% snap run --shell firefox
$ echo FILE:/tmp/krb5cc_746401104_OOmWAg 
FILE:/var/lib/snapd/hostfs//tmp/krb5cc_746401104_OOmWAg
```

The ticket was successfully used by Firefox to authenticate against a IIS server configured with the "Negotiate(Kerberos)" provider.

nteodosio commented:
Not really, as /tmp/krb5cc/x would pass it but would nonetheless not be
supported, so we need to test that /tmp is the directory component of the path.

Am 10/06/2025 um 03:49 schrieb ***@***.***:
> @bboozzoo commented on this pull request.
> 
> 
> 
> > @@ -760,6 +760,26 @@ func snapdHelperPath(toolName string) (string, error)
> > {  
>  	return filepath.Join(snapBase, dirs.CoreLibExecDir, toolName), nil
>  }
>  
> +/* Kerberos tickets live in /tmp, or in the place specified by KRB5CCNAME
> + * environment variable. However, in snaps /tmp is private.
> + * So rewire the environment variable to point to /var/lib/snapd/hostfs/tmp,
> + * in case the variable is unset or resolves tickets to /tmp/krb5cc*.
> + *
> + * Snaps that want to read Kerberos tickets must then connect to the
> + * kerberos-tickets interface to have access to the tickets.
> + */
> +func exposeKerberosTickets(info *snap.Info) (string, error) {
> +	krb5EnvVar := osGetenv("KRB5CCNAME")
> +	if strings.HasPrefix(krb5EnvVar, "FILE:") {
> +		path := strings.TrimPrefix(krb5EnvVar, "FILE:")
> +		path = filepath.Clean(path)
> +		if filepath.Dir(path) == "/tmp" && strings.HasPrefix(path,
> "/tmp/krb5cc") {
> 
> isn't this sufficient?
> 
> ```suggestion
> 		if strings.HasPrefix(path, "/tmp/krb5cc") {
> ```
>

nteodosio commented on cmd/snap/cmd_run.go:
```diff
@@ -760,6 +760,26 @@ func snapdHelperPath(toolName string) (string, error) {
+	if strings.HasPrefix(krb5EnvVar, "FILE:") {
+		path := strings.TrimPrefix(krb5EnvVar, "FILE:")
+		path = filepath.Clean(path)
+		if filepath.Dir(path) == "/tmp" && strings.HasPrefix(path, "/tmp/krb5cc") {
```
I'm pushed the test in a separate commit, but `os.Stat(actualKrbTicketPath)` fails with

> stat /var/lib/snapd/hostfs/tmp/krb5cc_286515778: no such file or directory

Do I need to somehow connect the dummy snap to the new interface? I do not see the Xauthority one doing it though, so I figured I'd ask first. (:

bboozzoo reviewed (commented):
Do yo think you could add a spread test for it as well to check the access from within a snap? Feel free to reach out on MM if you need guidance

bboozzoo commented on tests/main/interfaces-kerberos-tickets/kerberos-tickets-consumer:
you need to move the files around a bit to get a dir/file structure like this:
```
tests/main/interfaces-kerberos-tickets/
     \- kerberos-tickets-consumer/
         \- meta/
             \- snap.yaml
         \- kerberos-tickets-consumer
```

nteodosio replied:
With
```
├── kerberos-tickets-consumer/
│   ├── bin/
│   │   └── kerberos-tickets-consumer*
│   └── meta/
│       └── snap.yaml
└── task.yaml
```

I got
```
[...]
+ echo '# tasks executed on system'
# tasks executed on system
+ '[' -f /var/tmp/snapd-tools/runtime-state/runs ']'
++ cat /var/tmp/snapd-tools/runtime-state/runs
+ echo 'EXECUTED_TESTS=garden:ubuntu-24.04-64:tests/main/interfaces-kerberos-tickets '
EXECUTED_TESTS=garden:ubuntu-24.04-64:tests/main/interfaces-kerberos-tickets
+ echo ''
-----
.
2025-06-26 12:16:14 Starting shell to debug...
garden:ubuntu-24.04-64 .../tests/main/interfaces-kerberos-tickets#
```
I don't see the `echo`s from the test in any of the log though, so not sure if it really ran the test. [Full log for reference](https://termbin.com/oxda).

bboozzoo commented on tests/main/interfaces-kerberos-tickets/kerberos-tickets-consumer/meta/snap.yaml:
```diff
@@ -0,0 +1,9 @@
+description: Tries to read and then write to a Kerberos ticket
+
+apps:
+  kerberos-ticekts-consumer:
```
```suggestion
  kerberos-tickets-consumer:
```

bboozzoo commented on tests/main/interfaces-kerberos-tickets/task.yaml:
```diff
@@ -0,0 +1,46 @@
+
+execute: |
+    echo "KRB5CCNAME was correctly overwritten in the snap's environment."
+    kerberos-tickets-consumer | MATCH "KRB5CCNAME:$EXPECTED_KR5CCNAME"
```
```
+ echo 'KRB5CCNAME was correctly overwritten in the snap'\''s environment.'
KRB5CCNAME was correctly overwritten in the snap's environment.
+ kerberos-tickets-consumer
+ MATCH KRB5CCNAME:FILE:/var/lib/snapd/hostfs/tmp/krb5cc_test
grep error: pattern not found, got:
KRB5CCNAME:FILE:/var/lib/snapd/hostfs//tmp/krb5cc_test
Could not read ticket /var/lib/snapd/hostfs//tmp/krb5cc_test
Could not write to ticket /var/lib/snapd/hostfs//tmp/krb5cc_test
```
one `/` too many in the output

bboozzoo commented:
Thanks for the updates. The tests look good. The following failure is expected:

```
 ++ curl -I -o /dev/null -s -w '%{http_code}\n' --connect-timeout 10 https://snapcraft.io/docs/xilinx-dma-interface
+ status_code=200
+ '[' 200 -ge 200 ']'
+ '[' 200 -lt 300 ']'
+ return 0
+ echo 'Interface xilinx-dma has webpage'
Interface xilinx-dma has webpage
+ '[' -n '' ']'
+ '[' -n ' kerberos-tickets' ']'
+ echo 'The following interfaces do not have a documentation webpage:  kerberos-tickets'
The following interfaces do not have a documentation webpage:  kerberos-tickets
+ test -z ''
+ test -z ' kerberos-tickets'
```

as there's no documentation page for this interface yet. The test is expecting it at https://snapcraft.io/docs/kerberos-tickets which you can create by adding a new interface page in the snapcraft forum. I think it needs to be linked to from some index page. @degville can you help with that?

bboozzoo commented on cmd/snap/cmd_run.go:
```diff
@@ -760,6 +760,29 @@ func snapdHelperPath(toolName string) (string, error) {
+		path := strings.TrimPrefix(krb5EnvVar, "FILE:")
+		path = filepath.Clean(path)
+		if filepath.Dir(path) == "/tmp" && strings.HasPrefix(path, "/tmp/krb5cc") {
+			return "FILE:/var/lib/snapd/hostfs/" + path, nil
```
this should fix duplicate `/`:
```suggestion
			// with vanilla config this ends up returning:
			// FILE:/var/lib/snapd/hostfs/tmp/krb5cc_test
			return "FILE:" + filepath.Join("/var/lib/snapd/hostfs/", path), nil
```

jslarraz reviewed (approved):
LGTM

zyga commented on cmd/snap/cmd_run.go:
```diff
@@ -760,6 +760,31 @@ func snapdHelperPath(toolName string) (string, error) {
+ * kerberos-tickets interface to have access to the tickets.
+ */
+func exposeKerberosTickets(info *snap.Info) (string, error) {
+	krb5EnvVar := osGetenv("KRB5CCNAME")
```
Nitpick, can you either now or in a follow-up move this to package level const. The name is so peculiar and it is used more than once.

zyga commented on tests/main/interfaces-kerberos-tickets/task.yaml:
```diff
@@ -0,0 +1,46 @@
+    if and only if the kerberos-ticket interface is connected,
+    and can never write to it.
+
+systems: [ubuntu-24.04-64]
```
Could we run this on a wider array of systems or did you want to limit testing time?

nteodosio replied:
I just wanted to limit testing time. Is it preferrable to remove this statement and let it run wherever?

zyga replied:
Yes

zyga reviewed (approved):
LGTM with a few questions and comments inline.

nteodosio commented:
Just for your information, Codecov complained so I extended cmd_run_test.go to also check for the case where no KRB5CCNAME is in the environment. And also extended the spread test's task.yaml to check for an unsupported KRB5CCNAME.

zyga commented on tests/main/interfaces-kerberos-tickets/kerberos-tickets-consumer/bin/kerberos-tickets-consumer:
```diff
@@ -0,0 +1,5 @@
+ticket=${KRB5CCNAME#FILE:}
+echo "KRB5CCNAME:$KRB5CCNAME"
+cat "$ticket" 2>/dev/null || echo "Could not read ticket $ticket"
+echo 8 2>/dev/null >>"$ticket" || echo "Could not write to ticket $ticket"
```
Ta, can you add a comment in case there's another iteration please.

zyga commented:
@nteodosio still LGTM. Anytime our CI is happy.

nteodosio commented:
Extended cmd_run_test.go again because Codecov decided to complain again.

nteodosio commented:
Codecov is still failing although every line introduced in the merge request is covered, so I will ignore it as some sort of bug.

ernestl commented:
Hi @nteodosio, please have a look at the failures on core: 
<img width="992" height="279" alt="image" src="https://github.com/user-attachments/assets/bac89a28-5243-4eb4-8abc-40b51e0ef826" />

nteodosio commented:
Hey @ernestl, sorry, didn't notice those, looked at non-core ones which were flaky.

I must say I do not know why the interface is missing for ubuntu-core. If something obvious is missing please let me know.

nteodosio commented:
@ernestl those are now fixed thanks to paride.

maykathm commented:
Hi @nteodosio, it looks like the new spread test, interfaces-kerberos-tickets failed on amazon, centos, and fedora:
```
-----
+ echo 'KRB5CCNAME was correctly overwritten in the snap'\''s environment.'
KRB5CCNAME was correctly overwritten in the snap's environment.
+ kerberos-tickets-consumer
+ MATCH KRB5CCNAME:FILE:/var/lib/snapd/hostfs/tmp/krb5cc_test
+ tr -s /
+ echo 'The interface is disconnected by default'
The interface is disconnected by default
+ snap interfaces -i kerberos-tickets
+ MATCH -- '^- +kerberos-tickets-consumer:kerberos-tickets'

'snap interfaces' is deprecated; use 'snap connections'.
+ echo 'When the plug is connected'
When the plug is connected
+ snap connect kerberos-tickets-consumer:kerberos-tickets
+ echo 'Then the snap command can rean read the Kerberos ticket, but not write to it'
Then the snap command can rean read the Kerberos ticket, but not write to it
+ kerberos-tickets-consumer
+ MATCH contents
+ kerberos-tickets-consumer
+ MATCH 'cannot write to ticket'
grep error: pattern not found, got:
KRB5CCNAME:FILE:/var/lib/snapd/hostfs/tmp/krb5cc_test
contents
8
8
-----
```

nteodosio commented:
Thanks for raising that, Katie.

The test is showing that the snap is able to write to the ticket even before the interface is connected.

So the question is: Is Apparmor set up on those systems? I found at least one comment that says:

```
security-setuid-root/task.yaml-9-# No confinement (AppArmor, Seccomp) available on these systems
security-setuid-root/task.yaml:10:systems: [-debian-*, -fedora-*, -opensuse-*, -arch-*, -amazon-*, -centos-*]
```

but that is from 2018 and the test passed in Arch and Debian.

zyga commented on tests/main/interfaces-kerberos-tickets/task.yaml:
```diff
@@ -0,0 +1,49 @@
+    kerberos-tickets-consumer | MATCH "$TICKET_CONTENTS"
+    kerberos-tickets-consumer | MATCH "cannot write to ticket"
+
+    echo "When the plug is disconnected"
```
This must be guarded by "is apparmor effective":
```sh
if snap debug sandbox-features --required apparmor:kernel:file; then
   ...
fi
```

zyga commented on cmd/snap/cmd_run.go:
```diff
@@ -762,6 +762,31 @@ func snapdHelperPath(toolName string) (string, error) {
+	if strings.HasPrefix(krb5EnvVar, "FILE:") {
+		path := strings.TrimPrefix(krb5EnvVar, "FILE:")
+		path = filepath.Clean(path)
+		if filepath.Dir(path) == "/tmp" && strings.HasPrefix(path, "/tmp/krb5cc") {
```
The check is redundant, please keep only the `HasPrefix` part.

nteodosio replied:
It's not because without the first check `/tmp/krb5cc/a/b/c` would pass the test.

zyga reviewed (commented):
I missed a detail in tests that is sensitive to the presence of apparmor.

prottoyghose commented:
A quick thought/comment.

A few weeks back we started encountering the error `cmd_run.go:1408: WARNING: will not expose Kerberos tickets' path: Unsupported KRB5CCNAME: /home/username/krb5cc`. We started digging into it and encountered this PR and realized why this warning was popping up.

We have `firefox` installed via snap and it doesn't have access to `/tmp`. So we updated our `/etc/krb5.conf` file to have the ticket cache location set to user's home directory like `/home/username/krb5cc` and also export the environment variable `KRB5CCNAME` to `/home/username/krb5cc`. This fixed the firefox issues.

With these changes in PR, we realized it now expects `KRB5CCNAME` environment variable to be in a format with prefix being set to `FILE:` and also the ticket file to be in `tmp`. But I realized in order for this to work we still need to plug in the [Kerberos Interface](https://snapcraft.io/docs/kerberos-tickets-interface) which is still under development.

Should this change have been pushed before the interface was ready to prevent this warning? Some people may be confused that something is wrong with kerberos but it is just a warning since it can't find the ticket location in tmp or the `KRB5CCNAME` environment variable is now in an unsupported format.

nteodosio commented:
> A few weeks back we started encountering the error

It's not really an error, but a warning.

> We have `firefox` installed via snap and it doesn't have access to `/tmp`. So
> we updated our `/etc/krb5.conf` file to have the ticket cache location set to
> user's home directory like `/home/username/krb5cc` and also export the
> environment variable `KRB5CCNAME` to `/home/username/krb5cc`. This fixed the
> firefox issues.

That work-around is still working even with the warning, correct?

> With these changes in PR, we realized it now expects `KRB5CCNAME` environment
> variable to be in a format with prefix being set to `FILE:` and also the
> ticket file to be in `tmp`. But I realized in order for this to work we still
> need to plug in the [Kerberos
> Interface](https://snapcraft.io/docs/kerberos-tickets-interface) which is
> still under development.

The Kerberos interface is introduced by this very merge request and is released
in Snapd >= 2.71, but the Snap Store's review tools did not pick it up yet[1];
that's why you don't see it in Firefox. In any case, even if the interface were
connected, you will still get the warning as long as you don't want to mess
with your solution — indeed, why mess with what is not broken? (: — because the
path you are using is 'unexpected'.

> Should this change have been pushed before the interface was ready to prevent
> this warning? Some people may be confused that something is wrong with
> kerberos but it is just a warning since it can't find the ticket location in
> tmp or the `KRB5CCNAME` environment variable is now in an unsupported format.

It will only happen for users who deliberately altered KRB5CCNAME, so probably
they will realize that is OK. The warning is meant to help understanding the
issue in case people who are using this, as it were, default solution report
the issue persists.

Maybe we should then log this more inconspicuously instead of directly in the
stderr?

[1]https://launchpad.net/bugs/2123820

Am 18/09/2025 um 10:22 schrieb ***@***.***:
> prottoyghose left a comment (canonical/snapd#15519)
> 
> A quick thought/comment.
> 
> A few weeks back we started encountering the error `cmd_run.go:1408: WARNING:
> will not expose Kerberos tickets' path: Unsupported KRB5CCNAME:
> /home/username/krb5cc`. We started digging into it and encountered this PR
> and realized why this warning was popping up.
> 
> We have `firefox` installed via snap and it doesn't have access to `/tmp`. So
> we updated our `/etc/krb5.conf` file to have the ticket cache location set to
> user's home directory like `/home/username/krb5cc` and also export the
> environment variable `KRB5CCNAME` to `/home/username/krb5cc`. This fixed the
> firefox issues.
> 
> With these changes in PR, we realized it now expects `KRB5CCNAME` environment
> variable to be in a format with prefix being set to `FILE:` and also the
> ticket file to be in `tmp`. But I realized in order for this to work we still
> need to plug in the [Kerberos
> Interface](https://snapcraft.io/docs/kerberos-tickets-interface) which is
> still under development.
> 
> Should this change have been pushed before the interface was ready to prevent
> this warning? Some people may be confused that something is wrong with
> kerberos but it is just a warning since it can't find the ticket location in
> tmp or the `KRB5CCNAME` environment variable is now in an unsupported format.
>

prottoyghose commented:
@nteodosio yes, our current workaround is working even with the warning. We had to manually export the `KRB5CCNAME` variable so that it could be picked up by firefox and other chromium-based browsers that require kerberos ticket. Enhancing the log message would be nice since some people when see that warning they confuse it with an error and think that something is wrong with kerberos on their machine. either way once the interface is ready, we will revert our code since we don't want to be manually messing with this flow.

sviscaino commented:
Do snaps in `classic` confinement mode not have access to /tmp? We are running JetBrains PyCharm in classic, and it used to work fine with Kerberos, but the KRB5CCNAME in PyCharm now points to a non-existent file. After manually pointing that env var back to /tmp/krb5cc_{uid} in PyCharm, it works fine.

nteodosio commented:
Hi @sviscaino, that indeed was a detail missed but that Zygmunt caught and fixed in #15870; You should be able to revert to your old set-up at Snapd 2.72 (currently in edge and beta channels, you can try with `snap refresh --beta snapd`).
